### PR TITLE
Adds collect methods for arrays of results and maybes

### PR DESCRIPTION
**Pull request by JackSpagnoli**
Adds collect methods for arrays of results and maybes

Adds function `collect` as `Array<Result<O,E>> -> Result<Array<O>, Array<E>>` and `Array<Maybe<O>> -> Maybe<Array<O>>`.

For results, if any result in array is `Err`, returns an array of all `Err`s, else returns an array of all `Ok`s.
For maybes, if any result in array is `Nothing`, returns `Nothing`, else returns an array of all `Just`s.

Updates to version 2.3.0.

- [x] I have added or updated unit tests for this change.
- [x] I have updated the [changelog](https://github.com/rametta/pratica/blob/master/CHANGELOG.md) with info about the changes in this PR

### CODE REVIEW

**Comment by rametta** [comment]
Hello @JackSpagnoli, this is interesting. I like the idea, but I don't think we should be modifying the Array prototype. Would you be able to update and make it a utility like the [justs](https://github.com/rametta/pratica?tab=readme-ov-file#justs) function? Thanks!

File: src/collect.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      successes.push(result.value() as O)
+    } else {
+      failures.push(result.value() as E)
+    }
+  }
+
+  if (failures.length > 0) {
+    return Err(failures)
+  }
+  return Ok(successes)
```
Comment (rametta) [inline src/collect.ts:47]: Is there an existing `collect` function that you are modelling this logic off-of? I'm just curious how other libraries are using this

**Comment by JackSpagnoli** [comment]
Refactored to a utility function f4032cb9ae6313e7191b04e95c99b55e44d0bc84

File: src/collect.ts (hunk 1 above)
Comment (JackSpagnoli) [reply src/collect.ts:19]: I've done some digging, turns out I did something similar in Rust ages ago.
```
fn main() {
    let bad_values: Vec<Result<u32, &str>> = vec![Ok(1), Err("error"), Ok(3)];
    let bad_collection = bad_values.into_iter().collect::<Result<Vec<_>, _>>();

    assert!(bad_collection.is_err());
    assert_eq!(bad_collection.unwrap_err(), "error");

    let good_values: Vec<Result<u32, &str>> = vec![Ok(1), Ok(2), Ok(3)];
    let good_collection = good_values.into_iter().collect::<Result<Vec<_>, _>>();

    assert!(good_collection.is_ok());
    assert_eq!(good_collection.unwrap(), vec![1, 2, 3]);
}
```
That would explain why I thought to call it collect.

File: src/collect.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+export const collectResult = <O, E>(results: Array<Result<O, E>>): Result<Array<O>, Array<E>> => {
+  const successes: Array<O> = []
+  const failures: Array<E> = []
+
+  for (const result of results) {
+    if (result.isOk()) {
+      successes.push(result.value() as O)
+    } else {
+      failures.push(result.value() as E)
+    }
```
Comment (rametta) [inline src/collect.ts:13]: Would we be able to use `.cata()` here? Instead if having to do isOk and then type assertions. With cata the type should be built-in already

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,10 @@
 
 This document outlines the changes from version to version.
 
+## 2.3.0
+
+- Added `Array<Result>.collect()` and `Array<Maybe>.collect()`
```
Comment (rametta) [inline CHANGELOG.md:7]: Just need to update this again

File: src/collect.ts (hunk 2 above)
Comment (JackSpagnoli) [reply src/collect.ts]: 2336276e569d8268539fd44b8acc8a8c5cee2dc1

File: CHANGELOG.md (hunk 3 above)
Comment (JackSpagnoli) [reply CHANGELOG.md]: 794f605affeecd16590bed47f1b250d4ac91d860

**Review by rametta — APPROVED** [review]
Looks good, thanks!

**Pull request merged**